### ISSUE TITLE
Inconsistent Bucket Duration in Hammer for Periodic Rate Limits

### ISSUE THREAD

**Issue by vijayakumarkarthik** [ExHammer/hammer#103]
**Describe the bug**
In the current implementation of Hammer, the bucket duration can be less than the specified period depending on the precise moment a request is received. 
For example, with a 10-second period, requests made at 5 seconds past the epoch will result in a bucket duration of only 5 seconds instead of 10.

1. Configure Hammer with a period of 10 seconds. 
2. Make repeated requests at intervals that are not aligned with the period boundary. 
3. Observe that the bucket duration varies and does not consistently represent the full time period.

```elixir
defmodule MyAppWeb.TestRateLimiter do
  alias Hammer

  def test_rate_limit() do
    key = "test_key"
    limit = 1
    period = 10 * 1_000 # 10 seconds
    IO.inspect(Hammer.check_rate(key, limit, period))
    :timer.sleep(2_000)
    IO.inspect(Hammer.check_rate(key, limit, period))
    :timer.sleep(2_000)
    IO.inspect(Hammer.check_rate(key, limit, period))
  end 
end
```

** Provide the following details

- Elixir version (elixir -v): 1.17.2
- Erlang version (erl -v): 14.2.5.2
- Operating system: 

**Expected behavior**
When configuring a rate limit with a specified period (e.g., 10 seconds), each bucket should consistently cover the full time period. For example, a 10-second period should always represent a 10-second window, regardless of when the request is made within that period.

**Actual behavior**
The bucket duration can be less than the specified period. For example, with a 10-second period, requests made at 5 seconds past the epoch result in a bucket duration of only 5 seconds instead of 10. This leads to premature allowance of requests within the same time window.

```elixir
def stamp_key(id, scale_ms) do
  stamp = DateTime.to_unix(DateTime.utc_now(), :millisecond)
  bucket_number = div(stamp, scale_ms) * scale_ms
  key = {bucket_number, id}
  {stamp, key}
end
```

**Comment by epinault** [comment]
I ll see if I can look into it but PR welcome too if you have a solution

**Comment by ruslandoga** [comment]
👋 

I don't know about other backends, but this is unlikely to be implemented with ETS.

**Comment by epinault** [comment]
yea as I was looking back this is because it is a fixed window counter. I think this is fine to have it be the default. 
we could look at adding 

Sliding Window https://www.rdiachenko.com/posts/arch/rate-limiting/sliding-window-algorithm/ as well as Leaky Bucket support?

**Comment by ruslandoga** [comment]
Yeah, sliding window is easy to implement, here's a [clean example.](https://github.com/michalmuskala/plug_attack/blob/b3af6c1cd477872c5886d97c699a37fb7e638c63/lib/storage/ets.ex#L27-L44) For users with expensive resources, on which they rate limit to one per hour https://github.com/ExHammer/hammer/pull/99 it might be a very reasonable trade off.

**Comment by epinault** [comment]
7.x is almost ready. An rc2 is out with more type of throttling. The sliding window is implemented but as mentioned above, performance will be someting you need to consider as it is lot slower. But that will depend on the nature of your traffic and window size

**Comment by epinault** [comment]
This is likely fixed in 7.x
